File: common/core/src/commonMain/kotlin/com/trafi/ui/theme/Typography.kt
@file:Suppress("unused")

package com.trafi.ui.theme

expect val Inter: FontFamily

expect class FontFamily

expect class Sp(value: Float)

val Int.sp: Sp get() = Sp(this.toFloat())
val Float.sp: Sp get() = Sp(this)

expect enum class FontStyle {
    Normal
}
expect enum class FontWeight {
    Normal, Bold
}
expect class TextStyle(fontFamily: FontFamily? = null, fontStyle: FontStyle, fontWeight: FontWeight, fontSize: Sp, lineHeight: Sp) {
    val fontFamily: FontFamily?
    fun copy(fontFamily: FontFamily): TextStyle
}

data class MaasTypography internal constructor(
    val headingXXL: TextStyle,
    val headingXL: TextStyle,
    val headingL: TextStyle,
    val headingM: TextStyle,
    val textL: TextStyle,
    val textM: TextStyle,
    val textS: TextStyle,
    val label: TextStyle,
) {
    constructor(
        defaultFontFamily: FontFamily = Inter,
        headingXXL: TextStyle = TextStyle(
            fontStyle = FontStyle.Normal,
            fontWeight = FontWeight.Bold,
            fontSize = 40.sp,
            lineHeight = 50.sp
        ),
        headingXL: TextStyle = TextStyle(
            fontStyle = FontStyle.Normal,
            fontWeight = FontWeight.Bold,
            fontSize = 32.sp,
            lineHeight = 40.sp
        ),
        headingL: TextStyle = TextStyle(
            fontStyle = FontStyle.Normal,
            fontWeight = FontWeight.Bold,
            fontSize = 28.sp,
            lineHeight = 36.sp
        ),
        headingM: TextStyle = TextStyle(
            fontStyle = FontStyle.Normal,
            fontWeight = FontWeight.Bold,
            fontSize = 18.sp,
            lineHeight = 22.sp
        ),
        textL: TextStyle = TextStyle(
            fontStyle = FontStyle.Normal,
            fontWeight = FontWeight.Normal,
            fontSize = 16.sp,
            lineHeight = 24.sp
        ),
        textM: TextStyle = TextStyle(
            fontStyle = FontStyle.Normal,
            fontWeight = FontWeight.Normal,
            fontSize = 14.sp,
            lineHeight = 20.sp
        ),
        textS: TextStyle = TextStyle(
            fontStyle = FontStyle.Normal,
            fontWeight = FontWeight.Normal,
            fontSize = 12.sp,
            lineHeight = 16.sp
        ),
        label: TextStyle = TextStyle(
            fontStyle = FontStyle.Normal,
            fontWeight = FontWeight.Bold,
            fontSize = 14.sp,
            lineHeight = 18.sp
        ),
    ) : this(
        headingXXL = headingXXL.withDefaultFontFamily(defaultFontFamily),
        headingXL = headingXL.withDefaultFontFamily(defaultFontFamily),
        headingL = headingL.withDefaultFontFamily(defaultFontFamily),
        headingM = headingM.withDefaultFontFamily(defaultFontFamily),
        textL = textL.withDefaultFontFamily(defaultFontFamily),
        textM = textM.withDefaultFontFamily(defaultFontFamily),
        textS = textS.withDefaultFontFamily(defaultFontFamily),
        label = label.withDefaultFontFamily(defaultFontFamily),
    )
}

private fun TextStyle.withDefaultFontFamily(default: FontFamily): TextStyle {
    return if (fontFamily != null) this else copy(fontFamily = default)
}

File: ios/MaasCore/Sources/MaasTheme/Theme/Typography.swift
struct HeadingXXL: EnvironmentKey {
    static var defaultValue: TextStyle { TypographyScale().headingXXL }
}
struct HeadingXL: EnvironmentKey {
    static var defaultValue: TextStyle { TypographyScale().headingXL }
}
struct HeadingL: EnvironmentKey {
    static var defaultValue: TextStyle { TypographyScale().headingL }
}
struct HeadingM: EnvironmentKey {
    static var defaultValue: TextStyle { TypographyScale().headingM }
}
struct TextL: EnvironmentKey {
    static var defaultValue: TextStyle { TypographyScale().textL }
}
struct TextM: EnvironmentKey {
    static var defaultValue: TextStyle { TypographyScale().textM }
}
struct TextS: EnvironmentKey {
    static var defaultValue: TextStyle { TypographyScale().textS }
}
struct Label: EnvironmentKey {
    static var defaultValue: TextStyle { TypographyScale().label }
}
struct TextFont: EnvironmentKey {
    let get: (CGFloat) -> Font
    static var defaultValue: TextFont { TextFont { Font.system(size: $0) } }
}


public extension EnvironmentValues {

    var textStyleHeadingXXL: TextStyle {
        get { self[HeadingXXL.self].scaled(sizeCategory) }
        set { self[HeadingXXL.self] = newValue }
    }

    var textStyleHeadingXL: TextStyle {
        get { self[HeadingXL.self].scaled(sizeCategory) }
        set { self[HeadingXL.self] = newValue }
    }

    var textStyleHeadingL: TextStyle {
        get { self[HeadingL.self].scaled(sizeCategory) }
        set { self[HeadingL.self] = newValue }
    }

    var textStyleHeadingM: TextStyle {
        get { self[HeadingM.self].scaled(sizeCategory) }
        set { self[HeadingM.self] = newValue }
    }

    var textStyleTextL: TextStyle {
        get { self[TextL.self].scaled(sizeCategory) }
        set { self[TextL.self] = newValue }
    }

    var textStyleTextM: TextStyle {
        get { self[TextM.self].scaled(sizeCategory) }
        set { self[TextM.self] = newValue }
    }

    var textStyleTextS: TextStyle {
        get { self[TextS.self].scaled(sizeCategory) }
        set { self[TextS.self] = newValue }
    }

    var textStyleLabel: TextStyle {
        get { self[Label.self].scaled(sizeCategory) }
        set { self[Label.self] = newValue }
    }

    var textFont: (CGFloat) -> Font {
        get { self[TextFont.self].get }
        set { self[TextFont.self] = TextFont(get: newValue) }
    }
}

// MARK: - TextStyle

public extension View {

    func textStyle(_ textStyle: TextStyle) -> some View {
        ModifiedContent(content: self, modifier: TextStyleModifier(textStyle: textStyle))
    }
}

private struct TextStyleModifier: ViewModifier {

    let textStyle: TextStyle
    @Environment(\.textFont) var textFont

    func body(content: Content) -> some View {

        let textStyle = Kotlin(self.textStyle)

        let styled = content
            .lineSpacing(textStyle.lineSpacing)
            .font(
                textFont(textStyle.fontSize)
                    .weight(textStyle.fontWeight)
                    .style(textStyle.fontStyle)
            )


        if let color = textStyle.color {
            return AnyView(styled.foregroundColor(color))
        } else {
            return AnyView(styled)
        }
    }
}

private extension Font {

    func style(_ style: BasicFontStyle) -> Font {
        switch style {
        case .italic: return italic()
        case .normal: return self
        default: fatalError("Invalid 'BasicFontStyle' - \(self)")
        }
    }
}

extension BasicFontWeight {

    var fontWeight: Font.Weight {
        switch self {
        case .normal: return .regular
        case .semibold: return .semibold
        case .bold: return .bold
        default: fatalError("Invalid 'BasicFontWeight' - \(self)")
        }
    }
}

// MARK: - Scaling

extension TextStyle {

    func scaled(_ sizeCategory: ContentSizeCategory) -> TextStyle {

        let textStyle: UIFont.TextStyle

        switch fontSize {
        case 28...: textStyle = .title1
        case 22...: textStyle = .title2
        case 20...: textStyle = .title3
        case 17...: textStyle = .body
        case 16...: textStyle = .callout
        case 15...: textStyle = .subheadline
        case 13...: textStyle = .footnote
        case 12...: textStyle = .caption1
        default:    textStyle = .caption2
        }

        return self.withSize(CGFloat(fontSize) * textStyle.scaleFactor(for: sizeCategory))
    }

    func withSize(_ newFontSize: CGFloat) -> TextStyle {
        TextStyle(
            fontWeight: fontWeight,
            fontStyle: fontStyle,
            fontSize: Int32(newFontSize),
            lineSpacing: lineSpacing,
            color: color
        )
    }
}

private extension UIFont.TextStyle {

    func scaleFactor(for sizeCategory: ContentSizeCategory) -> CGFloat {

        let medium = UIFont.preferredFont(
            forTextStyle: self,
            compatibleWith: UITraitCollection(preferredContentSizeCategory: .medium)
        )

        let scaled = UIFont.preferredFont(
            forTextStyle: self,
            compatibleWith: UITraitCollection(preferredContentSizeCategory: sizeCategory.ui)
        )

        return scaled.pointSize / medium.pointSize
    }
}

private extension ContentSizeCategory {
    var ui: UIContentSizeCategory {
        switch self {
        case .extraSmall: return .extraSmall
        case .small: return .small
        case .medium: return .medium
        case .large: return .large
        case .extraLarge: return .extraLarge
        case .extraExtraLarge: return .extraExtraLarge
        case .extraExtraExtraLarge: return .extraExtraExtraLarge
        case .accessibilityMedium: return .accessibilityMedium
        case .accessibilityLarge: return .accessibilityLarge
        case .accessibilityExtraLarge: return .accessibilityExtraLarge
        case .accessibilityExtraExtraLarge: return .accessibilityExtraExtraLarge
        case .accessibilityExtraExtraExtraLarge: return .accessibilityExtraExtraExtraLarge
        }
    }
}

